### ISSUE TITLE
Create our V1 data store schema

### ISSUE THREAD

**Issue by zalun** [mozilla/prowac#1]
I've been thinking about that (writing some dry code)
There are at least 3 ways to represent our data (essentially redis db is represented as a top-level JSON):
1. JSON of current snapshot

```
{ 
  snapshotKey: JSON.stringify(fetchedDataFromTopSites),
  currentSnapshot: snapshotKey
}
```

That doesn't allow us to go deep to dynamically show data for each site separately
1. Object of JSON records:

```
{ 
  snapshotKey: {
    'someUrl': JSON.stringify(siteData),
    // 1M records ...
  },
  currentSnapshot: snapshotKey
}
```

This usage allows for dynamic browsing the site information (JSON.parse is needed on one record only), but fails if we want to analyze the data directly from database. If we would decide to save fetched data to database and do the analysis after all sites are fetched and saved, we would have to get entire list from db then do a lot of JSON.parse. Then we've got 2 options:
1. push the Object to a big array in memory and do calculations later on
2. create arrays with data prepared for analysis
1. Array of hashtags with unpacked information:

```
{
  `http://somesite': {
    'manifest': '...',
    'httpredirect': true
    // other fields ...
  },
  // 1M records ...
  currentSites: ['http://somesite', 1M items ...],
  snapshotKey: JSON.stringify(fetchedDataFromTopSites)
}
```

This solution might be easier to work on (programatically), as we may create a query for all httpredirects, etc. This however requires a bunch of multiple redis commands to be run in one call, and I've been reading that Redis has issues if there is 100000+. We can split it into 20 bunches, but still it is a pain...

My choice would be the second attempt, but we need to think if relational database wouldn't be a better option.

**Comment by TimAbraldes** [comment]
Thanks for looking into this @zalun!! Your analysis looks great and I appreciate you writing it up here.

Just to get this all straight in my head, I'll start by listing what I believe our database requirements are.

I believe that our database requirements from the dashboard are:
- Retrieve aggregate info (e.g. % of sites using Manifest) from the last complete crawl
- Retrieve historical aggregate info from each of the last 12 (or so) complete crawls
- Retrieve current/historical data about a specific URL

And that our database requirements from the crawler are:
- Retrieve the next URL to fetch/process for the current crawl
- Store the processed information for a specific URL for the current crawl (ideally atomic with next step)
- Remove URL from the list of URLs to fetch/process for the current crawl (ideally atomic with prev step)
- When the list of URLs to fetch/process is empty, aggregate all info from the current crawl and store the aggregated data

I like the "unpacked information" design that you suggested. I'm envisioning something like this:

``` json
{
  "historical-data": {
    "a-url.com": [
      {
        "timestamp": "a JS Date here",
        "hasProbe1": true,
        "hasProbe2": false,
        "hasProbe3": true
      },

      {
        "timestamp": "another JS Date",
        "hasProbe1": false,
        "hasProbe2": false,
        "hasProbe3": false
      }
    ],

    "another-url.com": [
      {
        "timestamp": "a JS Date here",
        "hasProbe1": false,
        "hasProbe2": false,
        "hasProbe3": true
      },

      {
        "timestamp": "another JS Date",
        "hasProbe1": false,
        "hasProbe2": false,
        "hasProbe3": true 
      }
    ],

    "a-third-url.com": [
      {
        "timestamp": "a JS Date here",
        "hasProbe1": true,
        "hasProbe2": true,
        "hasProbe3": true
      },

      {
        "timestamp": "another JS Date",
        "hasProbe1": false,
        "hasProbe2": false,
        "hasProbe3": false
      }
    ],

    "a-fourth-url.com": [
      {
        "timestamp": "a JS Date here",
        "hasProbe1": true,
        "hasProbe2": true,
        "hasProbe3": false
      },

      {
        "timestamp": "another JS Date",
        "hasProbe1": false,
        "hasProbe2": true,
        "hasProbe3": false
      }
    ]
  },
  "current-crawl-results": [
    {
      "a-url.com": {
        "timestamp": "a JS Date here",
        "hasProbe1": true,
        "hasProbe2": false,
        "hasProbe3": true
      },
      "another-url.com": {
        "timestamp": "a JS Date here",
        "hasProbe1": false,
        "hasProbe2": false,
        "hasProbe3": true
      }
    }
  ],
  "current-crawl-remaining-urls": [
      "a-third-url.com",
      "a-fourth-url.com"
  ],
  "aggregated-info": [
    {
      "date-aggregated": "a JS Date",
      "totalSites": 1000000,
      "haveProbe1": 2351,
      "haveProbe2": 71925,
      "haveProbe3": 10234
    },

    {
      "date-aggregated": "a JS Date",
      "totalSites": 1000000,
      "haveProbe1": 1234,
      "haveProbe2": 60124,
      "haveProbe3": 9123
    }
  ]
}
```

The crawler can spin through 'current-crawl-remaining-urls', processing each url and appending the result in 'current-crawl-results'. When 'current-crawl-remaining-urls' is empty, the crawler can aggregate the data in 'current-crawl-results', append the new aggregated data in 'aggregated-info', then it can empty 'current-crawl-results' into the appropriate URL-specific children of 'historical-data'.

This seems to exhibit the "bunch of multiple redis commands to be run in one call" issue that you mentioned, but only when the crawler is finished with a crawl and is putting the info from the last crawl into 'historical-data'. That should only happen a couple times per month I believe.

From my limited knowledge of Redis and its data types, I'm not sure if this layout is possible. What do you think? Is it possible to create something like this? Does this design do a good enough job of incorporating your analysis from above?

Thanks again!

**Comment by zalun** [comment]
As for multiple commands to database, we can add data one by one - without making it multiple. Just add some data after every crawled site. Collecting aggregated info is fine, as we can do this every time.

I don't see the reason to store unpacked data in database. There is no need to ever get all of historical data at once from database. Storing object data in key-value store is complicated. In an unpacked (not a JSON string) form we can store only hashes (shallow - one level only - Objects) and lists (Arrays). To achieve such construct we would need:

``` javascript
{
  // set of all urls ever appeared in top sites, might not be needed, but necessary 
  // if we would like to display records
  'historical_data': ['a-url-com', 'another-url-com', 'a-third-url-com'],
  // for each url a set of all dates
  'a-url-com': ['a-js-date-here', 'another-js-date'],
  // finally for each url and date combination an object
  'a-url-com-a-js-date-here': {
     "timestamp": "a JS Date here",
     "hasProbe1": true,
     "hasProbe2": false,
     "hasProbe3": true
  },
  'a-url-com-another-js-date': {
     "timestamp": "another JS Date",
     "hasProbe1": false,
     "hasProbe2": false,
     "hasProbe3": false
  }
}
```

And so on. It's not very complicated but requires going back and forth to database.

Summing, I think we don't need relational database (we can do heavy relations in Redis, but it's painful). BTW - my knowledge of Redis is also very limited, I haven't done any production site using it (my main db was always MySQL and Postgress). First living creature is Mercurius and then Platatus. I just seen what can be done and I like this as we can create sort of relational structure without hardcoding the schema.

[NOTE] I've added a section of stringified objects, but then I realized it requires the same number of Redis commands as the solution above

**Comment by digitarald** [comment]
I like @zalun's proposal with the relationship of sites and their timestamped crawler results. Building a key for `site-timestamp` is straightforward as it just concats the 2 fields. It is also _easy_ to get the last crawl date for one or all website and query the the result with those.

Some more considerations that came to mind:

For the dashboard we'll need do some more querying, probably with data in memory; which might overwhelm a single machine in the cloud. Over time the schema might advance to hold more pre-processed data (like set of sites that support service worker).

To save memory (and costs), we might want to shorten the keys by hashing them. Shortening a SHA-1 to the first 5 characters might be enough to avoid collisions.

**Comment by TimAbraldes** [comment]
After a bunch more reading about the Redis data types, I understand your points, @zalun, about using JSON strings as opposed to trying to store unpacked data in the DB.

My current understanding of how we want to design the DB is as follows:
- For each URL we've crawled, a key "historical-data:the-url.com" whose value is a Redis List of JSON strings. Each JSON string represents the processed data for that site from a particular crawl.
- A Redis List at the key "current-crawl-remaining-urls" containing 1 Redis String per URL remaining
- A Redis Hash at the key "current-crawl-aggregated-data" containing the aggregated data from the current crawl (e.g. number of sites with manifest)
- A key "aggregated-info" whose value is a Redis List of JSON strings. Each JSON string represents the aggregate data from a particular crawl.

I'll try to represent an example DB here:

```
// This is a Redis List
"current-crawl-remaining-urls": [
  "a-third-url.com",
  "a-fourth-url.com"
],

// This is a Redis Hash
"current-crawl-aggregated-data": {
  "startTime": "A JS Date/time",
  "totalSites": 467102,
  "haveProbe1": 2351,
  "haveProbe2": 71925,
  "haveProbe3": 10234
},

// This is a Redis List and each Redis String value that it
// contains is a JSON string
"aggregated-info": [
  JSON.stringify({
    "startTime": "A JS Date/time",
    "endTime": "A JS Date/time",
    "totalSites": 1000000,
    "haveProbe1": 2351,
    "haveProbe2": 71925,
    "haveProbe3": 10234
  }),

  JSON.stringify({
    "startTime": "A JS Date/time",
    "endTime": "A JS Date/time",
    "totalSites": 1000000,
    "haveProbe1": 1234,
    "haveProbe2": 60124,
    "haveProbe3": 9123
  })
],

// There will be one key like this for every URL we've ever
// crawled. The name is always "historical-data:some-url.com"
// and the value is always a Redis List containing JSON
// strings. The JSON strings represent the information from
// a particular crawl.
"historical-data:a-url.com": [
  JSON.stringify({
    "timestamp": "a JS Date here",
    "hasProbe1": true,
    "hasProbe2": false,
    "hasProbe3": true
  }),
  JSON.stringify({
    "timestamp": "another JS Date",
    "hasProbe1": false,
    "hasProbe2": false,
    "hasProbe3": false
  })
],

"historical-data:another-url.com": [
  JSON.stringify({
    "timestamp": "a JS Date here",
    "hasProbe1": false,
    "hasProbe2": false,
    "hasProbe3": true
  }),
  JSON.stringify({
    "timestamp": "another JS Date",
    "hasProbe1": false,
    "hasProbe2": false,
    "hasProbe3": true 
  })
],

"historical-data:a-third-url.com": [
  JSON.stringify({
    "timestamp": "a JS Date here",
    "hasProbe1": true,
    "hasProbe2": true,
    "hasProbe3": true
  }),
  JSON.stringify({
    "timestamp": "another JS Date",
    "hasProbe1": false,
    "hasProbe2": false,
    "hasProbe3": false
  })
],

"historical-data:a-fourth-url.com": [
  JSON.stringify({
    "timestamp": "a JS Date here",
    "hasProbe1": true,
    "hasProbe2": true,
    "hasProbe3": false
  }),
  JSON.stringify({
    "timestamp": "another JS Date",
    "hasProbe1": false,
    "hasProbe2": true,
    "hasProbe3": false
  })
]
```

Now for the types of queries we need to run against the DB:
- Retrieve aggregate info (e.g. % of sites using Manifest) from the last complete crawl

```
LRANGE aggregated-info 0 1
JSON.parse the result
```
- Retrieve historical aggregate info from each of the last 12 (or so) complete crawls

```
LRANGE aggregated-info 0 12
JSON.parse the result
```
- Retrieve current/historical data about a specific URL

```
LRANGE historical-data:a-url.com 0 12
JSON.parse the result
```
- Retrieve/remove the next URL to fetch/process for the current crawl

```
LPOP current-crawl-remaining-urls
```
- Store the processed information for a specific URL for the current crawl

```
LPUSH historical-data:a-url.com JSON.stringify(aResultsObject)
// Do the following only for probes that returned true
HINCRBY current-crawl-aggregated-data haveProbe1 1
HINCRBY current-crawl-aggregated-data haveProbe3 1
```
- When the list of URLs to fetch/process is empty, store the aggregated data for this now-completed crawl

```
HGETALL current-crawl-aggregated-data
// Parse the returned info into aggrObject
LPUSH aggregated-info JSON.stringify(aggrObject)
HMSET current-crawl-aggregated-data haveProbe1 0 haveProbe2 0 haveProbe3 0 ...
```

I believe that, for every type of query we want to run against our DB, we are performing only very fast and small-memory-footprint operations: We never have to loop over 1M+ records. Additionally, we can store an arbitrarily-large number of crawl results since we just use LRANGE to grab the last 12 (or 20, or however many we want).

I'm afraid I don't fully grasp the site-timestamp relationship suggestion. Could someone explain how that would change the schema as described here? Or if it would change the schema entirely, could we create a comment with the full schema and maybe what the queries would look like as well?

I think we're close to nailing down our V1 schema!

**Comment by zalun** [comment]
Now that's a comment! :)
Anyway - the only difference from your idea is that I wanted to be able to provide a list of timestamps and in here we would need to collect all data to get that, but on the other hand front-end (or API) might be better with some data instead of just dates. This structure is perfectly fine and I'm sure we can go with that.

**Comment by TimAbraldes** [comment]
closing this since it technically landed in #12. We can modify the data store in follow-up issues
